catalyst/callbacks/metric.py

```diff
@@ -1,66 +1,132 @@
 # @TODO: add metric aggregation, etc callback
+from abc import abstractmethod, ABC
+from typing import Dict, Iterable, Union, Tuple
+
+import torch
+
 from catalyst.core.callback import Callback, CallbackNode, CallbackOrder
 from catalyst.core.runner import IRunner
-from catalyst.metrics._metric import ICallbackBatchMetric, ICallbackLoaderMetric
+from catalyst.metrics import ICallbackBatchMetric, ICallbackLoaderMetric, IMetric
 
 
-class IMetricCallback(Callback):
+class IMetricCallback(Callback, ABC):
     """Metric callback interface, abstraction over metric step."""
+    @abstractmethod
+    def on_loader_start(self, runner: "IRunner") -> None:
+        pass
+
+    @abstractmethod
+    def on_batch_end(self, runner: "IRunner") -> None:
+        pass
 
-    pass
+    @abstractmethod
+    def on_loader_end(self, runner: "IRunner") -> None:
+        pass
 
 
-# @TODO: add KV support for input/output
-class BatchMetricCallback(IMetricCallback):
+class MetricCallback(IMetricCallback):
     def __init__(
-        self,
-        metric: ICallbackBatchMetric,
-        input_key: str,
-        target_key: str,
-        log_on_batch: bool = True,
+            self,
+            metric: Union[ICallbackBatchMetric, ICallbackLoaderMetric],
+            input_key: Union[str, Iterable[str], Dict[str, str]],
+            target_key: str,
```

Comment:
is it possible to support KV for `target_key` too? :)

Reply:
and work in 2 cases only:
- str for the input and str for the output
- key-value for the input and key-value for the output
?
what do u think about it?

Reply:
is such case it would be easier to redefine the `on_batch_end` logic during the `__init__` and reduce the number of if statements during the run